### PR TITLE
Jbaker/refactor viewable month range

### PR DESCRIPTION
**Pull request by jason-capsule42**
# Alaska Airlines Pull Request

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

**Resolves:** #213

## Summary:

This is a significant refactor of the logic controlling which and how many calendar months to render.

The new system follows theses rules:

### The first calendar month that "may" be rendered is chosen in the following priority order:

1. The `calendarStartDate` is set with a valid date
1. The `minDate` is set with a valid date

### The last calendar month that "may" be rendered is chosen in the following priority order:

1. The `calendarEndDate` is set with a valid date
1. The `maxDate` is set with a valid date

### The following logic is used for choosing which months are currently rendered in DOM

In mobile the full range is rendered from the first month that "may" be rendered (as determined above) to the last month that "may" be rendered. If no range is determined using the earliers rules, the mobile calendar will start from the current date and include the next 11 months (12 months total).

In desktop, 1 month will be rendered by default. If the `range` attribute is used, this will change to 2 months. If `calendarStartDate` is set, the first month rendered will be this date. Otherwise, if `value` is set, this date will be used. Otherwise, the current date will be used. 

The desktop calendar may navigate to previous and next months. The range of navigation is restricted by `calendarStartDate` and `calendarEndDate`. If one or the other is not set, `minDate` and `maxDate` are considered next. If none of these four attributes are set, the desktop calendar may be navigated forward and backward through time as far as the user wishes.

## Type of change:

Please delete options that are not relevant.

- [x] New capability
- [ ] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate)

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

File: src/auro-calendar.js [hunk 1]
```diff
@@ -66,103 +109,104 @@ export class AuroCalendar extends RangeDatepicker {
...
-    // if maxDate is defined and its earlier than the current month move the view back to the maxDate month
-    if (this.maxDate) {
-      const maxAsDateObj = new Date(this.maxDate);
-      const maxDateMonth = maxAsDateObj.getMonth() + 1;
-      const maxDateYear = maxAsDateObj.getFullYear();
-      const firstOfMonthForMaxDate = new Date(`${maxDateMonth}/01/${maxDateYear}`);
+    // Loop through the number of calendars to render and add the HTML
+    for (let cal = 0; cal < this.numCalendars; cal += 1) {
+      const date = this.firstRenderedMonth;
+      const newMonthDate = new Date(date.setMonth(date.getMonth() + 1));
```
Comment (jordanjones243) [inline src/auro-calendar.js:191]: Review logic

**Review by jordanjones243 — CHANGES_REQUESTED** [review]
December is jumping forward a year

**Review by jordanjones243 — COMMENTED** [review]
Calendar needs to auto scroll to `calendarFocusDate` in mobile

File: src/auro-datepicker.js [hunk 2]
```diff
@@ -11,6 +11,7 @@ import { LitElement } from "lit";
 import { html } from 'lit/static-html.js';
```
[the remark this answers is not in the document]
Comment (jordanjones243) [reply src/auro-datepicker.js:11]: Fixed

**Comment by jordanjones243** [comment]
> Calendar needs to auto scroll to `calendarFocusDate` in mobile

Error was resolved

File: src/auro-datepicker.js [hunk 3]
```diff
@@ -69,13 +73,21 @@ export class AuroDatePicker extends LitElement {
   constructor() {
     super();
 
+    /**
+     * @private
+     */
+    this.util = new AuroDatepickerUtilities();
```
Comment (blackfalcon) [inline src/auro-datepicker.js:79]: Consider grouping all your private defaults. See auro-icon as a good example. 

https://github.com/AlaskaAirlines/auro-icon/blob/6dcf1eb795cc7e032d713f050504eb51fc82c5b6/src/auro-icon.js#L52

File: src/utilitiesCalendarRender.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  determineNumCalendarsToRender(elem) {
+    // 1. Determine the maximum number of months that can be rendered.
+    //    This is based on the screen size and the defined range of months.
+    const maxRenderableMonths = this.maximumRenderableMonths(elem);
+
+    // 2. Start by assuming we can render the max number of months.
+    let calendarCount = maxRenderableMonths;
+
+    // 3. If we didn't get a count early, restrict based on min/max date.
+    if (!calendarCount && elem.minDate && elem.maxDate) {
```
[the remark this answers is not in the document]
Comment (blackfalcon) [reply src/utilitiesCalendarRender.js:75]: If `this.maximumRenderableMonths(elem)` returns a falsy value, `calendarCount` will also be falsy. Thus `undefined`. 

It's always good to have the check.

**Review by blackfalcon — CHANGES_REQUESTED** [review]

File: src/auro-datepicker.js (hunk 3 above)
Comment (jason-capsule42) [reply src/auro-datepicker.js:79]: I see your point. I took a slightly different approach - I put all the utility declarations first as I see them as something notably different than the other properties.

**Review by jordanjones243 — APPROVED** [review]

**Pull request merged**